docs: release blog for Ibis 9.0 (#8918)
ibis-project/ibis: docs/posts/ibis-version-9.0.0-release/index.qmd

---
title: "Ibis 9.0: SQLGlot-ification"
author: "Ibis team"
date: "2024-04-18"
categories:
    - release
    - blog
---

## Overview

> Ibis 9.0 completes ["the big

Reviewer: ```suggestion
Ibis 9.0 wraps up ["the big
```
Maybe? It just seems repetitive to say "completes" and then "completing" right after.

refactor"](../../posts/roadmap-2024-H1/index.qmd#the-big-refactor), completing
the transition from SQLAlchemy to [SQLGlot](https://github.com/tobymao/sqlglot)
and drastically simplifying the codebase. This is a big step toward stabilized
internals and allows us to more easily add new features and backends going
forward.

Ibis 9.0 also adds new functionality, fixes many bugs, and welcomes a new
committer to the project!

> ### New committer

Reviewer: @ncclementi is also a new committer: https://github.com/ibis-project/governance/discussions/6

Reviewer: I didn't realize this post was drafted well before that happened. Anyway.

Ibis now has a `describe` method to get a summary of a table:

```{python}
#| code-fold: true
#| echo: false
ibis.options.repr.interactive.max_rows = len(t.columns)
```

```{python}
t.describe()
```

```{python}
#| code-fold: true
#| echo: false
ibis.options.repr.interactive.max_rows = 3
```

### GraphViz custom attributes

Thanks to [a couple](https://github.com/ibis-project/ibis/pull/8510) [community
contributions](https://github.com/ibis-project/ibis/pull/8527), you can now pass
custom attributes to the GraphViz visualization of an expression.

:::{.callout-tip}
You can also call `expr.visualize()` and pass the same arguments.
:::

```{python}
from ibis.expr.visualize import to_graph

expr = (
    t.group_by("species", "island")
    .agg(count=t.count())
    .order_by(ibis._["count"].desc())
)

to_graph(
    expr,
    label_edges=True,
    node_attr={"shape": "hexagon", "color": "green", "fontname": "Roboto Mono"},
    edge_attr={"fontsize": "12", "fontname": "Comic Sans MS"},
)
```

### Polars input and output

You can now directly output a Polars dataframe from an Ibis table (instead of
going through Apache Arrow as previously possible):

```{python}
t.to_polars()
```

> And you can directly input a Polars dataframe into an Ibis memtable:

Reviewer: ```suggestion
And you can directly construct an Ibis memtable from a Polars dataframe:
```
Nit: Sounds more grammatically correct to me than inputting a dataframe?

```

### TopK includes NULLs

This could technically be considered a breaking change or bug fix. Regardless,
`topk` now includes NULLs by default:

```{python}
t["sex"].topk(3)
```

### Order and group by variadic arguments

For consistency with other methods, `order_by` and `group_by` now accept
variadic arguments:

```{python}
t.order_by(t["body_mass_g"].desc(), "island", t["species"])
```

```{python}
grouped = (
    t.group_by("species", "island")
    .agg(count=t.count())
    .order_by(ibis._["count"].desc())
)
grouped
```

You can still pass in a list or tuple if you prefer.

### Create more backends with empty URLs

For convenience, you can now create backends with an empty URL for Dask, pandas,
Polars, and DataFusion:

```{python}
con = ibis.connect("dask://")
con = ibis.connect("pandas://")
con = ibis.connect("polars://")
con = ibis.connect("datafusion://")
```

This is in addition to the existing backend-specific connection methods:

```{python}
con = ibis.dask.connect()
con = ibis.pandas.connect()
con = ibis.polars.connect()
con = ibis.datafusion.connect()
```

Note that URL parameters are not meaningful with these backends. The main use

> case is more convenient connection swapping programmatically:

Reviewer: What was the issue with `getattr(ibis, backend).connect()`? 🤷 I guess `getattr` feels more "advanced"?

Author: yeah nobody knows how to do that

```{python}
for backend in [
    "duckdb",
    "sqlite",
    "pyspark",
    "dask",
    "pandas",
    "polars",
    "datafusion",
]:
    con = ibis.connect(f"{backend}://")
    print(con)
```

## Wrapping up

Ibis 9.0 is an exciting release that completes the transition to SQLGlot, adds a

> lot of new functionality, and welcomes a new committer to the project!

Reviewer: ```suggestion
lot of new functionality, and welcomes two new committers to the project!
```

- saschahofmann

Thank you to everyone who contributed to this release! And a special thanks to
Krisztián Szűcs for his work on the internal representation and SQLGlot refactor
work, it has drastically improved the Ibis codebase.

With this refactor, it's never been easier to contribute to Ibis! Get on touch
via [GitHub](https://github.com/ibis-project/ibis) or
[Zulip](https://ibis-project.zulipchat.com/) if you're interested.

## Backends

No new backends were added in this release, [unless you count the April Fools'
Day Unix backend](../unix-backend/index.qmd)! As always, there have been
backend-specific improvements and bug fixes. Some highlights are below.

Check the [full changelog](../../release_notes.qmd#9.0.0) for more details.

### Snowflake

You can now run Ibis inside of a Snowflake stored procedure! Look for a blog on
this coming soon.

### DuckDB

In this release, `asof` joins are also now supported with a `tolerance`
parameter.

With [DuckDB's addition of MySQL as a database it can attach
to](https://duckdb.org/2024/01/26/multi-database-support-in-duckdb.html), we've
added a `read_mysql` function to Ibis inline with `read_postgres` and
`read_sqlite`.

> For geospatial operations, casting from binary to geometry is now supported.

Reviewer 2: adding something like: 
> This means that you can read a parquet file that has a geometry column in WKB format, and easily cast it to geometry file to do operations on it. 

pseudocode: 

```python 
table = ibis.read_parquet("your_parquet_file_with_wkb_geo_column")
table = table.mutate(geo_col=table.geometry.cast("geometry")) # table.geometry is the original column
```

Author: can we use `ibis.examples` for the data here?

Author: done w/ zones

Reviewer 2: @lostmygithubaccount Not really, the zones example is from a table in `geojson` meaning it's not in binary it already reads in the correct format. If want to read some light data you can use this file. 

https://github.com/ibis-project/testing-data/raw/master/parquet/geo_wkb.parquet

```python
t = ibis.read_parquet("https://github.com/ibis-project/testing-data/raw/master/parquet/geo_wkb.parquet")

t.geometry #you'll notice it's binary

t = t.mutate(geo_col=t.geometry.cast("geometry"))

t.geo_col.cantroid() #not you can do geo operations
```

:::

Ibis 9.0 also adds new functionality, fixes many bugs, and welcomes two new
committers to the project!

### Meet the Ibis team!

Members of the core Ibis team will be at [PyCon 2024 in Pittsburg, PA,
USA in May](https://us.pycon.org/2024/) and [SciPy 2024 in Tacoma, WA,
USA in July](https://www.scipy2024.scipy.org/)! Stop by at one of the talks or
tutorials to say hi in English and/or Español:

- [Tutorials: Introduction to Ibis: blazing fast analytics with DuckDB, Polars,
 Snowflake, and more, from the comfort of your Python
 repl.](https://us.pycon.org/2024/schedule/presentation/55/)
- [Charlas: Porque SQL está en todas
 partes.](https://us.pycon.org/2024/schedule/presentation/117/)

::: {.callout-note}
The schedule for SciPy 2024 is not available yet, keep an eye out for Ibis sessions.
:::

You can also listen to a [recent podcast episode with Real
Python](https://realpython.com/podcasts/rpp/201) where Phillip Cloud, lead
maintainer of Ibis, talks about the project.

More podcasts and talks are in the works, so stay tuned!

### New committers

We're excited to welcome [Nick Crews](https://github.com/NickCrews) and [Naty

> Clementi](https://github.com/ncclementi) as the newest committers to the Ibis

Reviewer 3: Maybe make `committers` a link to our governance doc, to indicate that here "committers" means a level of privileges rather than just "has made commits"?

### DuckDB

In this release, `asof` joins are also now supported with a `tolerance`
parameter.

With [DuckDB's addition of MySQL as a database it can attach
to](https://duckdb.org/2024/01/26/multi-database-support-in-duckdb.html), we've
added a `read_mysql` function to Ibis inline with `read_postgres` and
`read_sqlite`.

You can now cast binary data to geometry in the DuckDB backend, enabling use of regular Parquet files:

```{python}
zones = ibis.examples.zones.fetch().relocate("geom")  # <1>
zones.to_parquet("zones.parquet")  # <2>
```

1. Fetch the zones example data with the geometry column.
2. Write to a Parquet file, converting the geometry column to binary.

```{python}
zones = ibis.read_parquet("zones.parquet")  # <1>
zones.schema()  # <2>
```

1. Load in the zones Paruqet file.
2. Notice the geometry column is binary.

```{python}
zones = zones.mutate(geom=zones["geom"].cast("geometry"))  # <1>
zones.schema()  # <2>
```

1. Cast the binary geometry column to a geometry column.
2. Notice the geometry column is now a geometry type.

### PySpark

PySpark 3.5 is now supported.

### Exasol

Numerous improvements have been made to the Exasol backend by Nicola including

> bit operations, time operations, and hexdigest.

Reviewer 3: ```suggestion
bitwise operations, time operations, and hexdigest.
```

### Apache Flink

Scalar user-defined functions (UDFs) are now supported in the Flink backend.

### RisingWave

Streaming data definition language (DDL) support has been added to the
RisingWave backend.

## Breaking changes

You can view the [full changelog](../../release_notes.qmd#9.0.0) for additional
breaking changes. There is [one we expect to affect many
users](#what-does-schema-mean).

### What does "schema" mean?

In Ibis, a "schema" is a mapping of column names to their types:

```{python}
t.schema()
```

And as of Ibis 9.0, that is all "schema" means! In building a standard Python
dataframe interface, we must standardize naming conventions and terminology
across data systems. Many systems use the words "schema", "database", "catalog",

> and "namespace" to mean different things.

Reviewer 3: Maybe swap the "schema can mean other things" sentences and the "And as of ..." sentences. In this order it's a bit confusing if you're not familiar with other users of "schema".

Author: moved things around

> We have standardizes on `table` as the container of data which has a `schema`. A

Reviewer 3: The first sentence isn't additive and reads a bit odd IMO. How about.

```suggestion
Ibis now standardizes on "database" and "catalog". A
```

`database` is a collection of tables, and a `catalog` is a collection of
databases. Not all backends will support all levels of hierarchy.

> Using other `schema` methods as of Ibis 9.0 will give a deprecation warning, and

Reviewer 3: ```suggestion
Existing methods using the hierarchical meaning of "schema" will now raise a deprecation warning, and
```
